spacy/visualization.py

```diff
@@ -0,0 +1,601 @@
+            if whole_row_fg_color is not None:
+                fg_color = whole_row_fg_color
+            elif self.value_dependent_fg_colors is not None:
+                fg_color = self.value_dependent_fg_colors.get(value, None)
+            if fg_color is None and render_all_colors_within_values:
+                fg_color = self.fg_color
+            if self.value_dependent_bg_colors is not None:
+                bg_color = self.value_dependent_bg_colors.get(value, None)
+            if whole_row_bg_color is not None:
+                bg_color = whole_row_bg_color
+            elif bg_color is None and render_all_colors_within_values:
+                bg_color = self.bg_color
+        if fg_color is not None or bg_color is not None:
+            value = self.printer.text(value, color=fg_color, bg_color=bg_color)
+        return value + right_padding
+
+
+class Visualizer:
+    @staticmethod
+    def render_dependency_tree(sent: Span, root_right: bool) -> List[str]:
+        """
+        Returns an ASCII rendering of the document with a dependency tree for each sentence. The
+        dependency tree output for a given token has the same index within the output list of
+        strings as that token within the input document.
+
+        root_right: True if the tree should be rendered with the root on the right-hand side,
+                    False if the tree should be rendered with the root on the left-hand side.
+
+        Algorithm adapted from https://github.com/KoichiYasuoka/deplacy
+        """
+
+        # Check sent is really a sentence
+        assert sent.start == sent[0].sent.start
+        assert sent.end == sent[0].sent.end
+        heads: List[Optional[int]] = [
+            None
+            if token.dep_.lower() == "root" or token.head.i == token.i
+            else token.head.i - sent.start
+            for token in sent
+        ]
+        # Check there are no head references outside the sentence
+        assert (
+            len(
+                [
+                    head
+                    for head in heads
+                    if head is not None and (head < 0 or head > sent.end - sent.start)
+                ]
+            )
+            == 0
+        )
+        children_lists: List[List[int]] = [[] for _ in range(sent.end - sent.start)]
+        for child, head in enumerate(heads):
+            if head is not None:
+                children_lists[head].append(child)
+        all_indices_ordered_by_column: List[int] = []
+        # start with the root column
+        indices_in_current_column = [i for i, h in enumerate(heads) if h is None]
+        while len(indices_in_current_column) > 0:
+            assert (
+                len(
+                    [
+                        i
+                        for i in indices_in_current_column
+                        if i in all_indices_ordered_by_column
+                    ]
+                )
+                == 0
+            )
```

Comment:
I'm finding this pretty hard to read - is there a way to rewrite this so it won't be spread out as much?